src/storage/src/table/cell_based_table.rs

```diff
@@ -190,40 +200,41 @@ impl<S: StateStore, SER: RowSerializer> From<CellBasedTableBase<S, SER, READ_WRI
 }
 
 impl<S: StateStore, SER: RowSerializer, const T: AccessType> CellBasedTableBase<S, SER, T> {
+    pub(super) fn fallback_vnodes() -> Arc<Bitmap> {
+        lazy_static::lazy_static! {
+            static ref FALLBACK_VNODES: Arc<Bitmap> = {
+                let mut vnodes = BitmapBuilder::zeroed(VIRTUAL_NODE_COUNT);
+                vnodes.set(0, true);
+                vnodes.finish().into()
+            };
+        }
+        FALLBACK_VNODES.clone()
+    }
+
     fn new_inner(
         keyspace: Keyspace<S>,
         table_columns: Vec<ColumnDesc>,
         column_ids: Vec<ColumnId>,
         order_types: Vec<OrderType>,
         pk_indices: Vec<usize>,
-        #[allow(unused_assignments)] mut dist_key_indices: Option<Vec<usize>>,
+        dist_key_indices: Vec<usize>,
+        vnodes: Arc<Bitmap>,
     ) -> Self {
         let row_serializer = SER::create(&pk_indices, &table_columns, &column_ids);
 
         let mapping = ColumnDescMapping::new_partial(&table_columns, &column_ids);
         let schema = Schema::new(mapping.output_columns.iter().map(Into::into).collect());
         let pk_serializer = OrderedRowSerializer::new(order_types);
 
-        // TODO: enable this when we implement concat/merge iterators of different vnode prefix
-        dist_key_indices = None;
-        let dist_key_in_pk_indices = dist_key_indices.as_ref().map(|dist_key_indices| {
-            dist_key_indices
-                .iter()
-                .map(|&di| {
-                    pk_indices
-                        .iter()
-                        .position(|&pi| di == pi)
-                        .expect("distribution keys must be a subset of primary keys")
-                })
-                .collect_vec()
-        });
-
-        // TODO: allow the executor to specify the vnodes.
-        let vnodes = {
-            let mut b = BitmapBuilder::zeroed(VIRTUAL_NODE_COUNT);
-            b.set(DEFAULT_VNODE as usize, true);
-            b.finish()
-        };
+        let dist_key_in_pk_indices = dist_key_indices
+            .iter()
+            .map(|&di| {
+                pk_indices
+                    .iter()
+                    .position(|&pi| di == pi)
+                    .expect("distribution keys must be a subset of primary keys")
+            })
+            .collect_vec();
```

Comment:
Add some comments that this is check of dis key in pk_indices? I remembered we have similar code in Materialize Executor, which use `HashSet::is_subset` .

Reply:
This is also used for computing the indices of distribution keys in pk. You may find some explanations here.

https://github.com/singularity-data/risingwave/blob/d6ea0a49c1ea43d61eb4d4193ea8f31f7cf81ffb/src/storage/src/table/cell_based_table.rs#L103-L105